coalib/output/ConsoleInteraction.py

```diff
@@ -165,45 +165,35 @@ def acquire_actions_and_apply(console_printer,
     cli_actions = CLI_ACTIONS if cli_actions is None else cli_actions
     failed_actions = set()
     applied_actions = {}
+
     while True:
-        actions = []
+        action_dict = {}
+        metadata_list = []
+
         for action in cli_actions:
-            if action.is_applicable(result, file_dict, file_diff_dict) is True:
-                actions.append(action)
+            if action.is_applicable(result,
+                                    file_dict,
+                                    file_diff_dict,
+                                    tuple(applied_actions.keys())) is True:
+                metadata = action.get_metadata()
+                action_dict[metadata.name] = action
+                metadata_list.append(metadata)
 
-        if actions == []:
+        if not metadata_list:
             return
 
-        action_dict = {}
-        metadata_list = []
-        for action in actions:
-            metadata = action.get_metadata()
-            action_dict[metadata.name] = action
-            metadata_list.append(metadata)
-
         # User can always choose no action which is guaranteed to succeed
-        if apply_single:
-            ask_for_action_and_apply(console_printer,
-                                     section,
-                                     metadata_list,
-                                     action_dict,
-                                     failed_actions,
-                                     result,
-                                     file_diff_dict,
-                                     file_dict,
-                                     applied_actions,
-                                     apply_single=apply_single)
-            break
-        elif not ask_for_action_and_apply(console_printer,
-                                          section,
-                                          metadata_list,
-                                          action_dict,
-                                          failed_actions,
-                                          result,
-                                          file_diff_dict,
-                                          file_dict,
-                                          applied_actions,
-                                          apply_single=apply_single):
+        chosen_action = ask_for_action_and_apply(console_printer,
+                                                 section,
+                                                 metadata_list,
+                                                 action_dict,
+                                                 failed_actions,
+                                                 result,
+                                                 file_diff_dict,
+                                                 file_dict,
+                                                 applied_actions,
+                                                 apply_single=apply_single)
+        if not chosen_action or apply_single is True:
```

Comment:
@mrtes `apply_single` can never be `True`! Only `None`, `False`, or an action string, so remove `is True`

There need to be tests for this `apply_single` stuff...

Reply:
@mrtes Sorry! Just read the reason for `apply_single is True` in https://github.com/coala/coala/pull/4179#issuecomment-341969865 :)

Reply:
Yeah, that's what I thought, too.
However, apply_single is set to true in some of the tests for this function (which results in an endless loop when combined with my changes).

Therefore, I thought it would be sensible to catch that possibility.
I can remove it, but will have to change these tests as well.

The fact that this parameter can be a magic string or a boolean or none (in a language without union types) is questionable design and will likely lead to further bugs in the future.

Reply:
But it's still a bit strange... Before your changes, the `break` was always done `if apply_single`. Now `if apply_single != 'Do (N)othing'` (and not just `True`) then `chosen_action` will not be `False` and there will be no `break`. This is different behavior now

Reply:
You are right.
Possible fix: Always return false in the `if apply_single` branch of `ask_for_action_and_apply`.
In the current code this return value is not used `if apply_single`, so it is no change in behaviour.
Additionally, it would remove the `àpply_single is True` check.

Reply:
@mrtes Look at https://travis-ci.org/coala/coala/builds/300932914

I removed the `is True` and all tests are also passing. There need to be more specific tests for different `apply_single` states... and a better understanding of this former general `if apply_single: ... break`

@Nosferatul HELP! :D

Reply:
@mrtes Hadn't read your https://github.com/coala/coala/pull/4179#discussion_r150409810 before writing https://github.com/coala/coala/pull/4179#discussion_r150409980 :) But it seems we are getting to some common level of thinking about that weird stuff ;)

Reply:
@mrtes BTW: `chosen_action` is a misleading name for the result of `ask_for_action_and_apply`, since the value determines if there will be more choices afterwards

Reply:
@mrtes It's getting worse and worse :P I just realized that by setting `apply_single=True` the whole `test_acquire_actions_and_apply` becomes useless, because the simulated console inputs are never used. By `apply_single=True`, the whole console input is just bypassed. You can set `with simulate_console_inputs('to', 'whatever')` and the test passes

Reply:
@mrtes OMG :D `test_acquire_actions_and_apply` is defined a second time for testing `apply_single=True`, but without the same method name, so just replacing the original test

Let's quickly merge https://github.com/coala/coala/pull/4860 to get proper testing for this PR herre

Reply:
Woops, lots of stuff I didn't even think about.
And I also didn't really look into the existing tests besides getting them to run. 
The name `chosen_action` is a relict I forgot to adapt.
I would propose a refactoring of the `apply_single` stuff, after we're done here :)

Reply:
@mrtes Created the first issues for that refactoring as you can see below... So let's decide what to do before and what to do after merging this PR